ctapipe/core/qualityquery.py

```diff
@@ -147,10 +147,14 @@ def get_table_mask(self, table):
         n_criteria = len(self.criteria_names) + 1
         result = np.ones((n_criteria, len(table)), dtype=bool)
 
+        table["__index__"] = range(len(table))
+
+        by_id = table.group_by("tel_id")
```

Comment:
table.group_by makes a full copy of the input table. As provided in the astropy slack by tomaldcroft, the best solution seems create a new table only with the tel_ids and the index to use that for group_by:

```
from astropy.table import Table
import numpy as np

table = Table({
    "event_id": [1, 2, 3, 4, 5, 6, 7, 8],
    "tel_id": [1, 1, 2, 1, 2, 2, 1, 3],
    "value": np.random.normal(size=8),
})


def apply_mask(table):
    cuts = {1: 0.0, 2: 1.0, 3: 2.0}
    index_table = Table({"tel_id": table["tel_id"], "index": np.arange(len(table))})
    grouped = index_table.group_by("tel_id")
    del index_table

    table["mask"] = False

    for key, group_index in zip(grouped.groups.keys, grouped.groups):
        tel_id = key["tel_id"]
        index = group_index["index"]
        table["mask"][index] = table["value"][index] >= cuts[tel_id]


apply_mask(table)
print(table)

```